### ISSUE TITLE
Atomic `Polygon` model

### ISSUE THREAD

**Issue by amaury1093** [chronoscio/backend#53]
Proposed by @wtokumaru in #26:

> Can we separate the geojson data from the Territory and have it as a separate Polygon (or whatever) Entity that different territories can point to (like how they point to a Nation/PoliticalEntity)? We need this in order to avoid having like 20 Belaruses or Romes or Alsace-Loraines. Even the notion of a territory being something that trades hands or an administrative border/subdividision retained between social changes inherently can not be not limited to a single continuous time range...

This model would have the following fields:
- id
- name
- geometry (geojson field)
- references

I personally like this idea. I had in mind to create a new Territory each time a PoliticalEntity changes borders, thinking that some geojson will be duplicate, but it's okay b/c db storage is cheap. But if we want a clean solution, this is definitely the way to go.

We will create a small collection of all known atomic territories, and future mappers can just pick&assemble them to build Territories.

Note: This corresponds exactly to CIDOC's http://www.cidoc-crm.org/Entity/E94-Space-Primitive/Version-6.2.

---

Problems:

1. Once we have a small collection of Polygons, how to efficiently search through them?
    - We'll have a `name` field, but it can be duplicate, is not precise etc...
    - If the search algorithm is not performant, mappers will create duplicate polygons, and we'll have polygons that are 99% similar.
2. If I want to create the Territory of France today in 2018, probably I'll find the Polygons of the 101 departments in our collection. So UX-wise, I basically need to search/click on 101 items? (possibly more, if the department is not the most atomic polygon)

**Comment by dwaxe** [comment]
The database schema for this would look something like:

```
class GeoFeature(Model):
    name
    ...
    geom = GeometryField

class PoliticalEntity(Model):
    name
    ...
    owned_geometries = ManyToManyField(GeoFeature, through='OwnershipPeriod')

class OwnershipPeriod(Model):
    start_date
    end_date
    polygon = ForeignKey(GeoFeature)
    political_entity = ForeignKey(PoliticalEntity)
```

**Comment by whirish** [comment]
I like this idea, my only question would be how practical it would be. In cases where a SpacePrimitive is divided for a peace treaty, we would have to separate it into multiple SpacePrimitives, and then update all existing references to it. I suppose most land acquisitions aren't this arbitrary but I'm sure it's happened and could be a pain on the backend. 

> `owned_geometries = ManyToManyField(GeoFeature, through='OwnershipPeriod')`

I assume this is supposed to be `ManyToManyField(SpacePrimitive, through='OwnershipPeriod')`? Thanks for the reference schema.

**Comment by ataalik** [comment]
Well I guess you might be right. Only overhead work this would create for us making a system for somehow allowing editors to filter and chose from older borders. Conforming to CIDOC is also a plus.

**Comment by wtokumaru** [comment]
It should also be possible to just submit new borders alongside the set of existing ones that they select. Or to just submit entirely new borders as we currently do, in the same manner as we currently do.

**Comment by whirish** [comment]
I'm thinking of war occupations: very small portions of ground would be gained and lost by both sides on a daily basis, and an atomic polygon will be created for each. Existing references to larger polygons encompassing these areas that would have been owned by a nation will have to be updated to point to the new subdivided nations, which I suppose is not excessively difficult to implement though it could potentially be a huge database operation. My issue more lies with when mappers will have to create future territories; there will be massive amounts of incredibly small atomic polygons they will have to filter through and select to model a nation.

**Comment by amaurymartiny** [comment]
As @wtokumaru pointed out, I think that this solution is "at least as good, but most probably better" than the current implementation:
- right now we duplicate geojsons no matter what
- if we implement this, we might have duplicate atomic polygons. The better we make the search algorithm, the less we have duplicates. With some stitching cache proposed by @wtokumaru, it shouldn't even hurt query performance.
- On the long-term, it's easier to ask this data questions: "how many time has Alsace-Lorraine changed hands between 1850-1950"?

UX proposal: when the mapper creates the polygon for a political entity, he can hover his mouse over an empty area on the map, right-click and select "Search for territories here", and it would show on the left-pane all territories that covers that particular lat/lng point. The mapper can then just select the territory he wants to add to this political entity. Repeat until all the political entity is covered. tldr: I think we can find some UX solutions.

So overall, I'm in favor on implementing this.

**Comment by amaurymartiny** [comment]
Alsace changing hands - that's the best example where implementing non-overlapping SpacePrimitive would solve the whole problem: as you said we just change the color on the frontend, as the SpacePrimitive "Alsace" is already on the user's computer.

Now think about any territory invasion:
- you either fetch the whole nation's territory before invasion, and its territory after invasion (that's 2 times large network transfers)
- or you fetch the territory before invasion, and calculate the geometric diff (read: heavy operation) with the one after invasion, send only the diff to the frontend.
- or you fetch the territory before invasion (array of non-overlapping SpacePrimitives), calculate the *array* diff (read: simple operation) with the one after invasion, send only the diff to the frontend.

SpacePrimitive is the solution that allows:
- the less network transfer
- the less backend computation

tldr: it's the fastest.
